Feature/right click contextual open for ic popover menu

c95560 opened the pull request:
<!-- 🙏 Thank you for your contribution, it is greatly appreciated! -->

## Summary of the changes
Adds an optional right-click contextual menu open to the ic-popover-menu. This potentially reduces the number of mouse actions a user needs to take, and could provide useful mouse x-y position information for events (e.g. right-click - open menu - add new component at mouse x-y position)

## Related issue
None known

## Checklist
- [?] New web component and react stories added to demonstrate feature. Not clear if unit tests needed. Visual style not changed.
- [y] I have manually accessibility tested any changes, if relevant. Verified existing behaviour not broken.
- [y] I have ensured any changes match the Figma component library.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/mi6/ic-ui-kit?pullRequest=934) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/mi6/ic-ui-kit?pullRequest=934) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/mi6/ic-ui-kit?pullRequest=934) it.</sub>

jd239 commented:
Hi @c95560. Thank you for your contribution. Can you please refer to the [Contributing ](https://github.com/mi6/ic-ui-kit/blob/main/CONTRIBUTING.md) file and make updates to your files, rebase and commits were relevant.

GCHQ-Developer-530 reviewed (changes requested):
A couple of console logs left in there that can now be removed. Also some styling issues that should be easily fixed by running npm run prettier:fix 😊

GCHQ-Developer-741 committed:
style(web-components): updating border-radius of buttons using badges

added conditional class when button has a badge to change border radius to fit better with badge when focused

jd239 committed:
build(root): removed redundant npm scripts

Removed redundancy NPM scripts which are not required as part of development or build

jd239 committed:
build(web-components): removed redundant npm scripts

Removed redundant NPM scripts to avoid confusion with setting up for development or build

GCHQ-Developer-299 committed:
docs(react): update ic-navigation react story to have better labels

change navigation item labels from 'navigation' to unique words, for easier debug

GCHQ-Developer-299 committed:
fix(web-components): ic-button and ic-navigation no longer read twice to screen readers

aria-label set to null when icon-only buttons have a tooltip (which uses aria-labelledby) and
aria-label is set to null when navigation-item has a text label.

860

GCHQ-Developer-530 committed:
feat(react): add story for new sizes for checkboxes

Add story for new sizes for checkboxes

. #643

GCHQ-Developer-530 committed:
feat(web-components): add sizes small and large to checkboxes

Add sizes small and large to checkboxes with updates to chips and buttons to use new IcSizes type

. #643

GCHQ-Developer-530 committed:
chore(web-components): remove watch from storybook command

Remove watch from storybook command to fix error when building

GCHQ-Developer-530 committed:
feat(web-components): update iconography and add new colour variables

Update iconography to be more distinctive and add new colour variables for unknown status and
anomalous

. #713

GCHQ-Developer-530 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -37,6 +37,44 @@ export class PopoverMenu {
+   *  State for popover render position
+   */
+  @Prop() renderX: number = null;
+  @Prop() renderY: number = null;
```
This prop will need it's own description.

GCHQ-Developer-530 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -37,6 +37,44 @@ export class PopoverMenu {
+   * The ID of the element that can open the popover menu
+   * if mouseAnchor is set to true. If null, then use the root element.
+   */
+  @Prop() clickEventEle: string = document.documentElement.id;
+  private clickEventEleRect: DOMRect = undefined;
```
Please can you follow our [style guide](https://github.com/mi6/ic-ui-kit/blob/main/CONTRIBUTING.md#style-guide) and move the new private variables and props into the correct place in the file and in alphabetical order?

GCHQ-Developer-530 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -37,6 +37,44 @@ export class PopoverMenu {
+  private clickEventEleRect: DOMRect = undefined;
+
+  /**
+   *  State for popover render position
```
Think this prop description should be specific to renderX

GCHQ-Developer-530 reviewed (changes requested):
Couple more comments

GCHQ-Developer-530 reviewed (changes requested):
Please could you also squash your commits into one's that follow [conventional commits](https://github.com/mi6/ic-ui-kit/blob/main/CONTRIBUTING.md#git-commit). You should have one for your update to react stories and another for the web-components work you've done.

jd239 committed:
chore(web-components): reverting build script due to changes not updating

Removing the build watch npm script prevents any updates being made within Storybook while
developing. This has been reverted back.

c95560 commented:
@GCHQ-Developer-530

Noticed some issues resolving merge conflict - duplicated functions etc. 
Rebuilt branch enacting requested changes and retested.
- Fixed a console error bug related to bad anchor element DOM query.
- Added missing prop descriptions, e.g. for renderX
- Improved style by alphabetically ordering declarations of class props, functions etc
- recommitted code as a single commit via conventional commits plugin
- removed console debug messages that had been left in ic-popover-menu.tsx
- removed commented out debug code
- fixed style issues with npm run prittier:fix
- fixed (minor) merge conflicts with latest changes from main

GCHQ-Developer-530 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -137,17 +133,67 @@ export class PopoverMenu {
-    if (this.open && this.isNotPopoverMenuEl(ev)) {
-      // If menu is open and the next click on the document is not a popover El, close the popover
+  handleClick(ev: MouseEvent): void {
+    //document.documentElement.style.backgroundColor = 'red';
```
```suggestion
```

GCHQ-Developer-530 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -20,100 +20,96 @@ import { createPopper } from "@popperjs/core";
-        this.popoverMenuEls.unshift(this.backButton);
-      }
+  /**
+   * If true, the ic-popover-menu will open when a MouseEvent button press occurs.
```
```suggestion
   * If `true`, the ic-popover-menu will open when a MouseEvent button press occurs.
```

GCHQ-Developer-530 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -20,100 +20,96 @@ import { createPopper } from "@popperjs/core";
+   * The MouseEvent.button values that can trigger an ic-popover-menu open.
+   * Defaults to [2], i.e. mouse right click only
+   *
+   * 0 = left click
+   * 1 = middle click
+   * 2 = right click
+   * 3 = mouse browser back
+   * 4 = mouse browser forward
```
Might be worth exporting these as an enum and setting the type to that as we only want to accept these numbers, not any number, and they're for the equivalent close prop

GCHQ-Developer-530 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -20,100 +20,96 @@ import { createPopper } from "@popperjs/core";
-  @Prop() parentPopover?: HTMLIcPopoverMenuElement;
+  @Prop() renderX: number = null;
+  /**
+   *  absolute screen Y co-ordinate for popover menu render position.
```
```suggestion
   *  Absolute screen Y co-ordinate for popover menu render position.
```

GCHQ-Developer-530 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -20,100 +20,96 @@ import { createPopper } from "@popperjs/core";
 
   /**
-   * @internal The parent popover menu of a child popover menu.
+   *  absolute screen X co-ordinate for popover menu render position.
```
```suggestion
   *  Absolute screen X co-ordinate for popover menu render position.
```

GCHQ-Developer-530 commented:
Thanks for updating to our conventional commit style, there's just a couple of updates to be made with the commits:
- Remove the merge commit
- Separate work out into one commit containing only the react work -  this should contain the react scope instead of ic-popover-menu something like feat(react) and then detail the work you have done
- The rest of the work can go under the web-components scope - feat(web-components) and then detail the work you've done

c95560 commented:
@GCHQ-Developer-530 

All prior outstanding issues should be resolved.

- Resolved additional outstanding issues with style / commented code / linting issues
- Added enum to denote allowed mouse click actions for ic-popover-menu options
- Removed merge commit
- Synced with develop branch instead of main
- Separated commits to web component and react file changes

GCHQ-Developer-530 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -12,6 +12,14 @@ import {
 import { getSlotElements } from "../../utils/helpers";
 import { createPopper } from "@popperjs/core";
 
+enum IcPopoverMenuClickEnum {
+  Left = 0,
+  Middle = 1,
+  Right = 2,
+  Back = 3,
+  Forward = 4
+}
+
```
I think this should be exported from an ic-popover-menu.types.ts file like we've done for other components (see ic-button for an example)

GCHQ-Developer-530 commented:
Sorry, it seems to be failing on prettier. Please could you run prettier:fix?

c95560 committed:
feat(ic-popover-menu): options to anchor popover menu to cursor, and open/close via contextual click

Provides an option to open or close the ic-popover-menu menu with contextual mouse clicks on a
screen area denominated by a HTML element. The menu can optionally be anchored to the mouse cursor,
propagating screen x-y click event information when the menu is opened.

c95560 committed:
docs(packages/react/src/stories/ic-popover-menu.stories.mdx): new react story for mouse anchor feat

Adds a new react story demonstrating the ic-popover-menu mouse anchor and contextual click features
from e9a48da6ea09bddfa2659589d21b798a6bc2880a.

c95560 commented:
newest changes should rectify linting issues @GCHQ-Developer-530

GCHQ-Developer-530 reviewed (changes requested):
This change seems to have caused a unit test to fail, and isn't reaching the 80% test coverage threshold. Would you be able to add some tests for the code you've written? 😊
Also noticed there's a lot of console warnings about render X and render Y about them being immutable.

jd239 commented on packages/web-components/src/components/ic-popover-menu/ic-popover-menu.tsx:
```diff
@@ -20,100 +21,87 @@ import { createPopper } from "@popperjs/core";
+   * Defaults to `IcPopoverMenuClickEnum.Left` only.
    */
-  @Prop() submenuLevel: number = 1;
+  @Prop() mouseEventCloseButtonPressVals: IcPopoverMenuClickEnum[] = [
```
Can this be changed to `mouseButtonClosePopover` so it's more readable

jd239 reviewed (changes requested):
Apologies for the late response. First and foremost, thanks for contributing this complex functionality to the popover menu. 🎉 
A few things to amend in addition to the comments added:
- It's worth checking out the [Contribution file on git commits](https://github.com/mi6/ic-ui-kit/blob/main/CONTRIBUTING.md#git-commit) on how to format the commit messages so they pass the linting step during build.
- Rebase off the main repo to fix conflicts. 

Thanks 😃

c95560 commented:
I'm very short on time right now so probably won't have time to tidy this up to required standard if significant work is required.

Will revisit eventually - but more than happy to hand this off to another dev if they want to get it merged before any significant conflicts creep in.

GCHQ-Developer-465 commented:
> I'm very short on time right now so probably won't have time to tidy this up to required standard if significant work is required.
> 
> Will revisit eventually - but more than happy to hand this off to another dev if they want to get it merged before any significant conflicts creep in.

Hey there! Thank you for your contribution :) 

We are happy to carry on with this and update the PR. 

Thank you again, your contribution is much appreciated. We look forward to seeing more contributions!

jd239 commented:
This work is continues in https://github.com/mi6/ic-ui-kit/tree/feature/popover-right-click-functionality. This PR will be closed and a new PR will be created with the peer review comments.
